### PR TITLE
Add SHA-256 support for file verification

### LINKED ISSUE #855: Remote file verification hash support
Currently, [SHA-512 is the only supported hash function](https://github.com/coreos/ignition/blob/master/internal/util/verification.go#L67-L74) for remote file verification. 

Many resources (in my case, Docker Compose) are only released with SHA-256 sums, making verification currently impossible without generating my own compatible sums, which kind of defeats the purpose.

Is there a reason SHA-256 was excluded? If not, would it be possible to include it?

### PR DESCRIPTION
**Pull request by sohankunkerkar**
Fixes #855

### CODE REVIEW

**Comment by lucab** [comment]
ok to test

**Comment by lucab** [comment]
There are docs changes which should go together with this.
I know at least https://github.com/coreos/ignition/blob/master/doc/configuration-v3_1-experimental.md needs updating, but there may be more to be found by grepping through.

**Comment by jlebon** [comment]
Blackbox tests failing with:

```
    --- FAIL: TestIgnitionBlackBox/preemption.base.user.provider.config_ (4.72s)
        blackbox_test.go:121: preemption.base.user.provider.config 
        blackbox_test.go:258: Rendered Ignition Config:
            {
            			"ignition": {"version": "3.0.0"},
            			"storage": {
            				"files": [{
            					"path": "/ignition/provider",
            					"contents": {"source": "data:,p"},
            					"overwrite": true
            				}]}
            		}
        filesystem.go:154: ignition [-clear-cache -platform file -stage disks -root  -log-to-stdout --config-cache /var/tmp/ignition-blackbox-509204576/ignition.json]
        filesystem.go:159: PID: 20106
        filesystem.go:161: Ignition output:
            INFO     : Ignition v2.2.1-22-g7641317
            INFO     : Stage: disks
            ERROR    : unable to clear cache: remove /var/tmp/ignition-blackbox-509204576/ignition.json: no such file or directory
            INFO     : reading system config file "/var/tmp/ignition-blackbox-509204576/system/base.ign"
            DEBUG    : parsing config with SHA512: afa09ca2ed6bdb7ec4f14d9b03a638c75dd1be158054cfbb98190b8498c02f424e143973857791f0319980f529b43e2bc8edde2f6e745ee876b93640aee1b335
            WARNING  : warning at $.storage.files.0.mode, line 4 col 16: permissions unset, defaulting to 0644
            INFO     : reading system config file "/var/tmp/ignition-blackbox-509204576/system/user.ign"
            DEBUG    : parsing config with SHA512: 440d41f556bfe697644535e184a6cb45ef495f1fbb3a8240f9de08a7a072953b41adda987b514f64f12e9f82e6e2381f49b016187cfe815531f4f0c232e2d8a0
            WARNING  : warning at $.storage.files.0.mode, line 4 col 16: permissions unset, defaulting to 0644
            WARNING  : warning at $.storage.files.0.mode: permissions unset, defaulting to 0644
            INFO     : disks: disks passed
            INFO     : Ignition finished successfully
        filesystem.go:154: ignition [-clear-cache -platform file -stage mount -root /var/tmp/ignition-blackbox-509204576/hd0p9 -log-to-stdout --config-cache /var/tmp/ignition-blackbox-509204576/ignition.json]
        filesystem.go:159: PID: 20132
        filesystem.go:161: Ignition output:
            INFO     : Ignition v2.2.1-22-g7641317
            INFO     : Stage: mount
            INFO     : reading system config file "/var/tmp/ignition-blackbox-509204576/system/base.ign"
            DEBUG    : parsing config with SHA512: afa09ca2ed6bdb7ec4f14d9b03a638c75dd1be158054cfbb98190b8498c02f424e143973857791f0319980f529b43e2bc8edde2f6e745ee876b93640aee1b335
            WARNING  : warning at $.storage.files.0.mode, line 4 col 16: permissions unset, defaulting to 0644
            DEBUG    : parsed url from cmdline: ""
            INFO     : no config URL provided
            INFO     : reading system config file "/var/tmp/ignition-blackbox-509204576/system/user.ign"
            DEBUG    : parsing config with SHA512: 440d41f556bfe697644535e184a6cb45ef495f1fbb3a8240f9de08a7a072953b41adda987b514f64f12e9f82e6e2381f49b016187cfe815531f4f0c232e2d8a0
            WARNING  : warning at $.storage.files.0.mode, line 4 col 16: permissions unset, defaulting to 0644
            WARNING  : warning at $.storage.files.0.mode: permissions unset, defaulting to 0644
            INFO     : mount: mount passed
            INFO     : Ignition finished successfully
        filesystem.go:154: ignition [-clear-cache -platform file -stage files -root /var/tmp/ignition-blackbox-509204576/hd0p9 -log-to-stdout --config-cache /var/tmp/ignition-blackbox-509204576/ignition.json]
        filesystem.go:159: PID: 20139
        filesystem.go:161: Ignition output:
            INFO     : Ignition v2.2.1-22-g7641317
            INFO     : Stage: files
            INFO     : reading system config file "/var/tmp/ignition-blackbox-509204576/system/base.ign"
            DEBUG    : parsing config with SHA512: afa09ca2ed6bdb7ec4f14d9b03a638c75dd1be158054cfbb98190b8498c02f424e143973857791f0319980f529b43e2bc8edde2f6e745ee876b93640aee1b335
            WARNING  : warning at $.storage.files.0.mode, line 4 col 16: permissions unset, defaulting to 0644
            INFO     : using default filename
            INFO     : using config file at "config.ign"
            DEBUG    : parsing config with SHA512: f9dc049c76675bc32cc1c1487a4d4b3086cf2c05c4d7746357c3214a09bbe3307a4545f826e1ba8c2de69727615b59290250c9ec72903fef67fda35b67d3dba0
            WARNING  : warning at $.storage.files.0.mode, line 4 col 16: permissions unset, defaulting to 0644
            WARNING  : warning at $.storage.files.0.mode: permissions unset, defaulting to 0644
            DEBUG    : files: compiled without relabeling support, skipping
            INFO     : files: createFilesystemsFiles: createFiles: op(1): [started]  writing file "/var/tmp/ignition-blackbox-509204576/hd0p9/ignition/base"
            INFO     : files: createFilesystemsFiles: createFiles: op(1): [finished] writing file "/var/tmp/ignition-blackbox-509204576/hd0p9/ignition/base"
            INFO     : files: createFilesystemsFiles: createFiles: op(2): [started]  writing file "/var/tmp/ignition-blackbox-509204576/hd0p9/ignition/provider"
            INFO     : files: createFilesystemsFiles: createFiles: op(2): [finished] writing file "/var/tmp/ignition-blackbox-509204576/hd0p9/ignition/provider"
            INFO     : files: files passed
            INFO     : Ignition finished successfully
        filesystem.go:154: ignition [-clear-cache -platform file -stage umount -root /var/tmp/ignition-blackbox-509204576/hd0p9 -log-to-stdout --config-cache /var/tmp/ignition-blackbox-509204576/ignition.json]
        filesystem.go:159: PID: 20145
        filesystem.go:161: Ignition output:
            INFO     : Ignition v2.2.1-22-g7641317
            INFO     : Stage: umount
            INFO     : reading system config file "/var/tmp/ignition-blackbox-509204576/system/base.ign"
            DEBUG    : parsing config with SHA512: afa09ca2ed6bdb7ec4f14d9b03a638c75dd1be158054cfbb98190b8498c02f424e143973857791f0319980f529b43e2bc8edde2f6e745ee876b93640aee1b335
            WARNING  : warning at $.storage.files.0.mode, line 4 col 16: permissions unset, defaulting to 0644
            DEBUG    : parsed url from cmdline: ""
            INFO     : no config URL provided
            INFO     : reading system config file "/var/tmp/ignition-blackbox-509204576/system/user.ign"
            DEBUG    : parsing config with SHA512: 440d41f556bfe697644535e184a6cb45ef495f1fbb3a8240f9de08a7a072953b41adda987b514f64f12e9f82e6e2381f49b016187cfe815531f4f0c232e2d8a0
            WARNING  : warning at $.storage.files.0.mode, line 4 col 16: permissions unset, defaulting to 0644
            WARNING  : warning at $.storage.files.0.mode: permissions unset, defaulting to 0644
            INFO     : umount: umount passed
            INFO     : Ignition finished successfully
        validator.go:231: Contents of file /var/tmp/ignition-blackbox-509204576/hd0p9/ignition/user do not match! u unset
        validator.go:231: Contents of file /var/tmp/ignition-blackbox-509204576/hd0p9/ignition/provider do not match! unset p
```

You should also be able to run it locally.

(Checking right now how hard it'd be to move it over to CoreOS CI.)

**Comment by jlebon** [comment]
> (Checking right now how hard it'd be to move it over to CoreOS CI.)

OK, opened https://github.com/coreos/ignition/pull/967 for this!

File: config/v3_0/types/verification.go [hunk 1]
```diff
@@ -56,6 +56,8 @@ func (v Verification) Validate(c path.ContextPath) (r report.Report) {
 	switch function {
 	case "sha512":
 		hash = crypto.SHA512
+	case "sha256":
```
Comment (lucab) [inline config/v3_0/types/verification.go:59]: I think we only want to do this for the upcoming 3.1.0, without retro-actively affecting 3.0.0.

File: doc/examples.md [hunk 2]
```diff
@@ -106,7 +108,31 @@ This example Ignition configuration will locate the device with the "VAR" filesy
 }
 ```
 
-The SHA512 sum of the file can be determined using `sha512sum`.
```
Comment (lucab) [inline doc/examples.md:109]: Here and below, I think it is enough to mention that SHA256 is supported too.

File: doc/examples.md [hunk 3]
```diff
@@ -106,7 +106,7 @@ This example Ignition configuration will locate the device with the "VAR" filesy
...
+The SHA512 sum of the file can be determined using `sha512sum`. It also supports the SHA256 format which is determined using `sha256sum`.
```
Comment (bgilbert) [inline doc/examples.md:109]: `It also supports` has an unclear referent.  What about "SHA256 sums are also supported, and can be calculated using `sha256sum`."?

File: tests/negative/files/invalid_hash.go [hunk 4]
```diff
@@ -62,6 +64,44 @@ func InvalidHash() types.Test {
 	}
 }
 
+func InvalidHashForSHA256() types.Test {
+	name := "files.verification.badhash.dataurl.sha256"
+	in := types.GetBaseDisk()
+	out := in
+	mntDevices := []types.MntDevice{
+		{
```
Comment (bgilbert) [inline tests/negative/files/invalid_hash.go:72]: Not needed.

File: tests/negative/files/invalid_hash.go [hunk 5]
```diff
@@ -62,6 +64,44 @@ func InvalidHash() types.Test {
...
+				"format": "ext4",
+				"wipeFilesystem": true
+			}],
+			"files": [{
+				"path": "/tmp0/test",
+				"contents": {
+					"source": "data:,asdf", "verification": {"hash": "sha256-e57cc41647638ccb9fb6844f2810807bcaa2aa800438cfb065f17adf6afb48d0"}}
+			}]}
+	}`
+	configMinVersion := "3.0.0"
```
Comment (bgilbert) [inline tests/negative/files/invalid_hash.go:93]: Should be `3.1.0-experimental`.

File: tests/negative/files/invalid_hash.go [hunk 6]
```diff
@@ -62,6 +64,44 @@ func InvalidHash() types.Test {
...
+			Label:        "EFI-SYSTEM",
+			Substitution: "$DEVICE",
+		},
+	}
+
+	config := `{
+		"ignition": {"version": "$version" },
+		"storage": {
+			"filesystems": [{
+				"path": "/tmp0",
```
Comment (bgilbert) [inline tests/negative/files/invalid_hash.go:82]: Let's not mount an extra filesystem for this test.  It's not necessary here.

File: tests/negative/files/invalid_hash.go [hunk 7]
```diff
@@ -97,3 +137,39 @@ func InvalidHashFromHTTPURL() types.Test {
...
+	    "files": [{
+	      "path": "/foo/bar",
+	      "contents": {
+	        "source": "http://127.0.0.1:8080/contents",
+			"verification": {"hash": "sha256-352cb4e231c03f9941d54aeee7da755504a7f2096338c609ba5d1b82143419c6"}
+	      }
+	    }]
+	  }
+	}`
+	configMinVersion := "3.0.0"
```
Comment (bgilbert) [inline tests/negative/files/invalid_hash.go:157]: `3.1.0-experimental`

File: tests/negative/files/invalid_hash.go [hunk 8]
```diff
@@ -97,3 +137,39 @@ func InvalidHashFromHTTPURL() types.Test {
...
+	in := types.GetBaseDisk()
+	out := in
+	config := `{
+	  "ignition": { "version": "$version" },
+	  "storage": {
+	    "files": [{
+	      "path": "/foo/bar",
+	      "contents": {
+	        "source": "http://127.0.0.1:8080/contents",
+			"verification": {"hash": "sha256-352cb4e231c03f9941d54aeee7da755504a7f2096338c609ba5d1b82143419c6"}
```
Comment (bgilbert) [inline tests/negative/files/invalid_hash.go:152]: The hash of HTTP sources should be a constant in `servers`, same as for SHA512.

File: tests/negative/files/invalid_hash.go [hunk 9]
```diff
@@ -97,3 +137,39 @@ func InvalidHashFromHTTPURL() types.Test {
...
+	  }
+	}`
+	configMinVersion := "3.0.0"
+	out[0].Partitions.AddFiles("ROOT", []types.File{
+		{
+			Node: types.Node{
+				Name:      "bar",
+				Directory: "foo",
+			},
+			Contents: "asdf\nfdsa",
```
Comment (bgilbert) [inline tests/negative/files/invalid_hash.go:164]: It's a negative test, so we don't actually expect this file to show up.

File: tests/positive/files/hash.go [hunk 10]
```diff
@@ -61,6 +64,42 @@ func ValidateFileHashFromDataURL() types.Test {
...
+	out[0].Partitions.AddFiles("ROOT", []types.File{
+		{
+			Node: types.Node{
+				Name:      "bar",
+				Directory: "foo",
+			},
+			Contents: "example file\n",
+		},
+	})
+	configMinVersion := "3.0.0"
```
Comment (bgilbert) [inline tests/positive/files/hash.go:92]: This test doesn't actually succeed, does it?  It should only work on `3.1.0-experimental`.

File: tests/positive/files/hash.go [hunk 11]
```diff
@@ -97,6 +136,42 @@ func ValidateFileHashFromHTTPURL() types.Test {
...
+	out[0].Partitions.AddFiles("ROOT", []types.File{
+		{
+			Node: types.Node{
+				Name:      "bar",
+				Directory: "foo",
+			},
+			Contents: "asdf\nfdsa",
+		},
+	})
+	configMinVersion := "3.0.0"
```
Comment (bgilbert) [inline tests/positive/files/hash.go:164]: Same.

File: tests/positive/files/hash.go [hunk 12]
```diff
@@ -133,3 +208,40 @@ func ValidateFileHashFromHTTPURLUsingHeaders() types.Test {
 		ConfigMinVersion: configMinVersion,
 	}
 }
+
+func ValidateFileHashFromHTTPURLUsingHeadersForSHA256() types.Test {
```
Comment (bgilbert) [inline tests/positive/files/hash.go:212]: I don't think this test is necessary.  Let's drop it to save some runtime.

File: tests/positive/files/hash.go [hunk 13]
```diff
@@ -97,6 +135,42 @@ func ValidateFileHashFromHTTPURL() types.Test {
...
+	in := types.GetBaseDisk()
+	out := types.GetBaseDisk()
+	config := `{
+	  "ignition": { "version": "$version" },
+	  "storage": {
+	    "files": [{
+	      "path": "/foo/bar",
+	      "contents": {
+	        "source": "http://127.0.0.1:8080/contents",
+			"verification": {"hash": "sha256-e57cc41647638ccb9fb6844f2810807bcaa2aa800438cfb065f17adf6afb48d0"}
```
Comment (bgilbert) [inline tests/positive/files/hash.go:149]: The hash of HTTP sources should be a constant in `servers`, same as for SHA512.

File: tests/negative/files/invalid_hash.go (hunk 8 above)
Comment (bgilbert) [reply tests/negative/files/invalid_hash.go:139]: Whoops, this comment was meant to go on the positive test.

**Pull request merged**